tiledb/sm/filter/bit_width_reduction_filter.cc

```diff
@@ -0,0 +1,457 @@
+/**
+ * @file   bit_width_reduction_filter.cc
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2017-2018 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file defines class BitWidthReductionFilter.
+ */
+
+#include "tiledb/sm/filter/bit_width_reduction_filter.h"
+#include "bit_width_reduction_filter.h"
+#include "tiledb/sm/filter/filter_pipeline.h"
+#include "tiledb/sm/misc/logger.h"
+#include "tiledb/sm/misc/utils.h"
+#include "tiledb/sm/tile/tile.h"
+
+namespace tiledb {
+namespace sm {
+
+/** Compute the number of bits required to represent a signed integral value. */
+template <typename T>
+static inline uint8_t bits_required(T value, std::true_type) {
```

Comment:
Although probably not important, but this could be improved by specializing on `T`. Just noting it here in case the total time this function is calculated shows up in benchmarking.